python_ta/contracts/__init__.py

```diff
@@ -603,3 +583,37 @@ def _debug(msg: str) -> None:
         return
 
     print("[PyTA]", msg, file=sys.stderr)
+
+
+def _set_invariants(klass: type) -> None:
+    """Retrieve and set the representation invariants of this class"""
+    # Update representation invariants from this class' docstring and those of its superclasses.
+    rep_invariants: List[Tuple[str, CodeType]] = []
+
+    # Iterate over all inherited classes except builtins
+    for cls in reversed(klass.__mro__):
+        if "__representation_invariants__" in cls.__dict__:
+            rep_invariants.extend(cls.__representation_invariants__)
+        elif cls.__module__ != "builtins":
+            assertions = parse_assertions(cls, parse_token="Representation Invariant")
+            # Try compiling assertions
+            for assertion in assertions:
+                try:
+                    compiled = compile(assertion, "<string>", "eval")
+                except:
+                    _debug(
+                        f"Warning: representation invariant {assertion} could not be parsed as a valid Python expression"
+                    )
+                    continue
+                rep_invariants.append((assertion, compiled))
+
+    setattr(klass, "__representation_invariants__", rep_invariants)
+
+
+def check_invariants(obj: object) -> None:
+    """Check that the representation invariants of obj are satisfied."""
+    klass = obj.__class__
+    klass_mod = _get_module(klass)
+
+    _set_invariants(klass)
```

Comment:
Okay I thought about this and I want `check_invariants` to be purely non-mutating, and so it should only check invariants that have already been parsed/added, and *not* call `_set_invariants` here. So please remove this call, but you can keep the helper function refactoring you made.

This also has an implication for the test you added, since you'll need to add the `@check_contracts` decorator to your class. This means that you can't violate the representation invariant when you initially create the object (since an `AssertionError` will be raised), but instead will need to create a valid object then mutate it to violate an RI. I suspect doing something like `self.age -= ...` will work, but if not, you can create an attribute that's a mutable object and then *mutate* that attribute value's object to violate an RI, e.g. `self.something.append(...)`. PythonTA doesn't check RIs after one of its attribute objects is mutated.